src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -274,19 +318,42 @@ public bool TryGetSemanticModel([NotNullWhen(returnValue: true)] out SemanticMod
                     return null;
                 }
 
-                if (this.TryGetSemanticModel(out var semanticModel))
+                SemanticModel? semanticModel;
+                if (disableNullableAnalysis)
                 {
-                    return semanticModel;
+#pragma warning disable RSEXPERIMENTAL001 // Internal usage of experimental API
+                    if (this.TryGetNullableDisabledSemanticModel(out semanticModel))
+#pragma warning restore
+                    {
+                        return semanticModel;
+                    }
+                }
+                else
+                {
+                    if (this.TryGetSemanticModel(out semanticModel))
+                    {
+                        return semanticModel;
+                    }
                 }
 
                 var syntaxTree = await this.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
                 var compilation = await this.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
 
-                var result = compilation.GetSemanticModel(syntaxTree);
+#pragma warning disable RSEXPERIMENTAL001 // sym-shipped usage of experimental API
+                var result = compilation.GetSemanticModel(syntaxTree, disableNullableAnalysis ? SemanticModelOptions.DisableNullableAnalysis : SemanticModelOptions.None);
+#pragma warning restore RSEXPERIMENTAL001
                 Contract.ThrowIfNull(result);
+                WeakReference<SemanticModel>? original = null;
 
                 // first try set the cache if it has not been set
-                var original = Interlocked.CompareExchange(ref _model, new WeakReference<SemanticModel>(result), null);
+                if (disableNullableAnalysis)
+                {
+                    original = Interlocked.CompareExchange(ref _nullableDisabledModel, new WeakReference<SemanticModel>(result), null);
```

Comment:
Is it possible just to have a ref local point to either of the WeakReferences and reuse the selection for the earlier if/else?